packages/core/src/components/axes/tool-bar.ts

```diff
@@ -0,0 +1,325 @@
+		// listen to show-tooltip Custom Events to render the tooltip
+		this.services.events.addEventListener(
+			Events.Toolbar.SHOW_MENU_ITEMS,
+			() => {
+				this.overflowMenuOptions.html(this.getOverflowMenuHTML());
+			}
+		);
+
+		// listen to hide-tooltip Custom Events to hide the tooltip
+		this.services.events.addEventListener(
+			Events.Toolbar.HIDE_MENU_ITEMS,
+			() => {
+				this.overflowMenuOptions.html(null);
+			}
+		);
+	}
+
+	render(animate = true) {
+		const isDataLoading = this.zoomService.isDataLoading();
+		this.isZoomBarEnabled =
+			this.zoomService.isZoomBarEnabled() &&
+			!this.zoomService.isEmptyState();
+		// size of overflow menu icon with background
+		const iconSize = Configuration.toolBar.iconSize;
+
+		const svg = this.getContainerSVG();
+		const { width } = DOMUtils.getSVGElementSize(
+			this.services.domUtils.getMainSVG(),
+			{
+				useAttrs: true
+			}
+		);
+
+		// overflow menu option width is 160px
+		// it's set by Carbon component
+		const overflowMenuOptionWidth = 160;
+		// no good solution to get correct ToolBar position
+		// parent x doesn't work well
+		// assume the overflowMenu button has right alignment in layout
+		this.overflowMenuX = width - overflowMenuOptionWidth;
+		this.overflowMenuY =
+			parseFloat(this.parent.node().getAttribute("y")) + iconSize;
+
+		const container = DOMUtils.appendOrSelect(svg, "svg.toolbar-container")
+			.attr("width", "100%")
+			.attr("height", Configuration.toolBar.height)
+			.attr("opacity", 1);
+
+		// clean children first
+		container.html(null);
+		// loading or empty state
+		if (isDataLoading) {
+			// put an empty rect to keep space unchanged
+			DOMUtils.appendOrSelect(container, "svg.toolbar-loading-space")
+				.append("rect")
+				.attr("height", Configuration.toolBar.height)
+				.attr("width", iconSize * 3) // value doesn't matter but can't be empty
+				.attr("opacity", 0);
+		} else {
+			const self = this;
+
+			const buttonList = [];
+			// add zoom in/out button only if zoom bar is enabled
+			if (this.isZoomBarEnabled) {
+				buttonList.push(this.getZoomInButtonConfig());
+				buttonList.push(this.getZoomOutButtonConfig());
+			}
+			buttonList.push(this.getOverFlowMenuButtonConfig());
+
+			// render buttons sequentially
+			let buttonXPosition = 0;
+			buttonList.forEach((button) => {
+				// zoom in icon and event
+				const buttonContainer = DOMUtils.appendOrSelect(
+					container,
+					`svg.${button.id}`
+				);
+				// add icon back ground rect
+				let buttonSVG = `<rect class="icon-${button.iconId}" x="${buttonXPosition}px" y="0px"
+					width="${iconSize}px" height="${iconSize}px"/>`;
+				buttonSVG += button.iconSVG(buttonXPosition);
+				buttonContainer.html(buttonSVG);
+				buttonContainer.on("click", button.clickFunction);
+				buttonXPosition += iconSize;
+			});
+
+			this.overflowMenuIcon = DOMUtils.appendOrSelect(
+				this.getContainerSVG(),
+				"rect.icon-overflowRect"
+			);
+
+			const hasOpenedOverflowMenuOptions =
```

Comment:
`dropdownIsOpen`?

Reply:
we just need more clarity in variable names, in many places. Reading a variable name I need to immediately understand what it's telling me

Reply:
Updated in https://github.com/carbon-design-system/carbon-charts/pull/717/commits/a956942af3621b1929cddc8ff2092810d6ca3804.